firefox_puppeteer/ui/windows.py

```diff
@@ -0,0 +1,486 @@
+        """
+        self.switch_to(handle)
+
+        # TODO: Maybe needs to wait as handled via an observer
+        return self.marionette.close_chrome_window()
+
+    def close_all(self, exceptions=None):
+        """Closes all open chrome windows.
+
+        There is an optional `exceptions` list, which can be used to exclude
+        specific chrome windows from being closed.
+
+        :param exceptions: Optional, list or a single entry of handles or
+         :class:`BaseWindow` instances not to close
+        """
+        handles_to_keep = exceptions or []
+        if not isinstance(handles_to_keep, list):
+            handles_to_keep = [handles_to_keep]
+
+        # Ensure we only have handles and no BaseWindow entries
+        handles_to_keep = [entry.handle if isinstance(entry, BaseWindow) else entry
+                           for entry in handles_to_keep]
+
+        # Find handles to close and close them all
+        handles_to_close = set(self.marionette.chrome_window_handles) - set(handles_to_keep)
+        for handle in handles_to_close:
+            self.close(handle)
+
+    def create_window_instance(self, handle, expected_class=None):
+        """Creates a :class:`BaseWindow` instance for the given chrome window
+
+        :param handle: The handle of the chrome window
+        :param expected_class: Optional, check for the correct window class
+        """
+        current_handle = self.marionette.current_chrome_window_handle
+        window = None
+
+        try:
+            # Retrieve window type to determine the type of chrome window
+            if handle != self.marionette.current_chrome_window_handle:
+                self.switch_to(handle)
+            window_type = self.marionette.get_window_type()
+        finally:
+            # Ensure to switch back to the original window
+            if handle != current_handle:
+                self.marionette.switch_to_window(current_handle)
+
+        if window_type == 'navigator:browser':
+            window = BrowserWindow(lambda: self.marionette, handle)
+        else:
+            raise errors.UnknownWindowError('Unknown window type "%s" for handle: "%s"' %
+                                            (window_type, handle))
+
+        if expected_class is not None and type(window) is not expected_class:
+            raise errors.UnexpectedWindowTypeError('Expected window "%s" but got "%s"' %
+                                                   (expected_class, type(window)))
+
+        return window
+
+    def focus(self, handle):
+        """Focuses the chrome window with the given handle.
+
+        :param handle: The handle of the chrome window
+        """
+        self.switch_to(handle)
+
+        with self.marionette.using_context('chrome'):
+            self.marionette.execute_script(""" window.focus(); """)
+
+        wait = Wait(self.marionette, timeout=5)
+        wait.until(lambda m: handle == self.focused_chrome_window_handle)
```

Comment:
This is not true. It defaults to None:
https://dxr.mozilla.org/mozilla-central/source/testing/marionette/client/marionette/wait.py#30

But I see where the problem is, it needs an update of the documentation so it says 5s but not None. I assume this would be good for a first good bug.